### ISSUE TITLE
Too high memory usage when syncing many files

### ISSUE THREAD

**Issue by guruz** [owncloud/client#2929]
We need to verify this (massif tool?) but the current assumption is that the syncing with its new job-based architecture uses too much memory.

One needs to think about
* Peak usage during discovery
* Peak usage during sync

Some ideas:
* csync structures not cleared before propagate (has been done for 1.8beta2 though)

* Somehow don't create all all jobs at once
 * Do the full discovery, but only create x jobs and then restart sync or propagation?

* Make sure sync jobs are deleted when finished

* Slim down SyncFileItem
 * ```log``` structure?
 * Switching things to ```QByteArray```
 * Each job with a pointer to a central SyncFileItem in the engine instead of a copy? (Makes code more fragile though)

Some opinions/snippets:
```
<olivier> IMHO, the best way to reduce the memory usage is to reduce the size of SyncFileItem
```

```
<eagles0513875> guruz: jturcotte it seems like my client chokes really when it hits memory usage 800mb +
<eagles0513875> guruz: jturcotte i can confirm selective syncing a bit at a time is alot more stable then one bulk sync
```

Related https://github.com/owncloud/client/issues/2902

**Comment by guruz** [comment]
> Make sure sync jobs are deleted when finished

Assigned to @ogoffart

**Comment by eagles051387** [comment]
I am the same eagles0513875 btw. How come sync doesn't behave like that of drop box or Google drive as soon as files are added an instant sync takes place and then once sync is complete the memory that it used to determine whats already been synced in that directory is freed. Right now upon sync completion memory usage does not go back down as it seems like memory isn't being freed after a complete sync

**Comment by guruz** [comment]
A bit of work has been done by @jturcotte in https://github.com/owncloud/client/commit/db7919dc2f0c90e95c8fc317d1104771a1d8f55e

@eagles051387 You can re-test with the nightly build that will be produced in 18h :)
http://download.owncloud.com/desktop/daily/

**Comment by eagles051387** [comment]
no problem Will test I dunno if you guys want to generate an installer for
me to test before then I can get it stress testing as we speak.

On Wed, Mar 11, 2015 at 1:59 PM, Markus Goetz <notifications@github.com>
wrote:

> A bit of work has been done by @jturcotte <https://github.com/jturcotte>
> in db7919d
> <https://github.com/owncloud/client/commit/db7919dc2f0c90e95c8fc317d1104771a1d8f55e>
>
> @eagles051387 <https://github.com/eagles051387> You can re-test with the
> nightly build that will be produced in 18h :)
> http://download.owncloud.com/desktop/daily/
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/owncloud/client/issues/2929#issuecomment-78257902>.
>



-- 
Jonathan Aquilina

**Comment by guruz** [comment]
@eagles051387 You can try ownCloud-1.8.0.4826-nightly20150311-setup.exe .. freshly generated. But I think it doesn't change enough memory usage yet..

**Comment by eagles051387** [comment]
Ok will wait for the build and install it tomorrow morning.

On Wed, Mar 11, 2015 at 3:43 PM, Markus Goetz <notifications@github.com>
wrote:

> @eagles051387 <https://github.com/eagles051387> You can try
> ownCloud-1.8.0.4826-nightly20150311-setup.exe .. freshly generated. But I
> think it doesn't change enough memory usage yet..
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/owncloud/client/issues/2929#issuecomment-78276311>.
>



-- 
Jonathan Aquilina

**Comment by guruz** [comment]
The build is already there http://download.owncloud.com/desktop/daily/ownCloud-1.8.0.4826-nightly20150311-setup.exe

**Comment by eagles051387** [comment]
Wow my bad been a busy chaotic day for me will test once I get home.

On Wed, Mar 11, 2015 at 3:49 PM, Markus Goetz <notifications@github.com>
wrote:

> The build is already there
> http://download.owncloud.com/desktop/daily/ownCloud-1.8.0.4826-nightly20150311-setup.exe
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/owncloud/client/issues/2929#issuecomment-78277607>.
>



-- 
Jonathan Aquilina

**Comment by dragotin** [comment]
@eagles051387 how does it look like with final 1.8.0 ?

**Comment by eagles051387** [comment]
Its better but still needs work. before it crashes it gets to 950mb of ram
usage and sometimes 1.1gb.

The problem lies in discovery phase. each chunk that gets done discovery
eats up and consumes more memory then previously reason being is its
reindexing those files that were already synced.

What I woudl suggest if files have already been synced they get checked but
memory as soon as its known that the file is on the remote machine the
memory gets freed.

On Wed, Mar 18, 2015 at 12:11 PM, Klaas Freitag <notifications@github.com>
wrote:

> @eagles051387 <https://github.com/eagles051387> how does it look like
> with final 1.8.0 ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/owncloud/client/issues/2929#issuecomment-82906363>.
>



-- 
Jonathan Aquilina

**Comment by eagles051387** [comment]
Have any further improvements been made that need testing out?

On Wed, Mar 18, 2015 at 12:41 PM, Jonathan Aquilina <eagles051387@gmail.com>
wrote:

> Its better but still needs work. before it crashes it gets to 950mb of ram
> usage and sometimes 1.1gb.
>
> The problem lies in discovery phase. each chunk that gets done discovery
> eats up and consumes more memory then previously reason being is its
> reindexing those files that were already synced.
>
> What I woudl suggest if files have already been synced they get checked
> but memory as soon as its known that the file is on the remote machine the
> memory gets freed.
>
> On Wed, Mar 18, 2015 at 12:11 PM, Klaas Freitag <notifications@github.com>
> wrote:
>
>> @eagles051387 <https://github.com/eagles051387> how does it look like
>> with final 1.8.0 ?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/owncloud/client/issues/2929#issuecomment-82906363>.
>>
>
>
>
> --
> Jonathan Aquilina
>



-- 
Jonathan Aquilina

**Comment by guruz** [comment]
Related https://github.com/owncloud/client/issues/3064 https://github.com/owncloud/client/issues/1817

**Comment by eagles051387** [comment]
The only logical thing I can think of is there is 1) a race condition some
where which is randomly triggered or 2) a potential memory leak, but
problem with the memory leak idea is that most likely that would surface
across all platforms no?

On Fri, Apr 17, 2015 at 9:59 AM, Markus Goetz <notifications@github.com>
wrote:

> Related #3064 <https://github.com/owncloud/client/issues/3064> #1817
> <https://github.com/owncloud/client/issues/1817>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/owncloud/client/issues/2929#issuecomment-93938285>.
>



-- 
Jonathan Aquilina

**Comment by guruz** [comment]
Linking commit (1.9) here: https://github.com/owncloud/client/commit/12d7484b85c92a0c897daf37fa992a330fb91d9f

**Comment by Mikhail-I** [comment]
1.8.1 windows client. Memory consumption is 500Mb after several hours of work and never drops down. Memory leak?

**Comment by jturcotte** [comment]
After investigation it seems like the actual memory issue we were tracking was caused by a flood of requests which was tracked by #3122. This is on top of the work done for #2902.

The memory usage can currently reach 500MB for a sync involving 100k files, but it should only be happening on the first sync and I think that this is acceptable (even though not ideal). Since this issue is too generic, please open a new issue if you have a more specific memory usage under a specific scenario.

@Mikhail-I It's a memory leak only if it continues growing (e.g. past 1GB). The memory can't go down easily due to heap fragmentation. Please open a new issue if you have more specific steps to reproduce the issue.

**Comment by eagles051387** [comment]
I think the biggest question is if this is related to what i was seeing in
a way. Do you have windows integration enabled?

On Wed, Jun 3, 2015 at 5:02 PM, Jocelyn Turcotte <notifications@github.com>
wrote:

> Closed #2929 <https://github.com/owncloud/client/issues/2929>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/owncloud/client/issues/2929#event-321390650>.
>



-- 
Jonathan Aquilina

**Comment by johnsonreyliu** [comment]
my mac github desktop goes all the way up to 30gb at some instances crashing everything!

**Comment by guruz** [comment]
@johnsonreyliu This is the repository for the **ownCloud** client, please try https://github.com/contact for github support
